### DISCUSSION TITLE
Dynamic Prompts/Wildcard in Adetailer extremely important

### DISCUSSION THREAD

**Discussion by Woisek** [adieyal/sd-dynamic-prompts#738] [Ideas]
With the increasing importance of using Adetailer, I think it would be beneficial for all, if defined and/or random generated prompts are re-usable in the Adetailer prompts, especially when we want something 'random specific'.

'Random specific' means for me, a random generated face in the main prompt, that should be enhanced in Adetailer. I create my random faces via full name (first and last name) to make it unique and reproducible.

Right now, the Adetailer prompts are 'dead' to wildcards, and variables that are created in the main prompt. I think it would be extremely important if Adetailer would be included in this process so we have complete control over our randomly generated prompts.

**Comment by uorufu** [comment]
I am no programmer, but I maneged to make this fork of wildcards:
https://github.com/uorufu/stable-diffusion-webui-wildcards-adetailer

It works well with adetailer. No matter what order you put your wildcards in or which you want to also use in adetailer, it will not generate different results unless the seed changes. So every seed is determined.

I changed a few functionalities, with my fork you can tier wildcards. Example prompt:  

` __1_lowerbody__ __1_upperbody__ __1_face1__ __2_face2__ `
would mean that if you make sure lower and upper body txt files have the same amount of lines, that you can match clothing properly while only taking a part of it to adetailer for face or breasts inpainting.
So you can put the 2 faces in face adetailer prompt and upper body in breast detailer prompt, etc. 

Order also doesn't matter.


At the moment there are only ten tiers. ( __0_txt__ --> __9_txt__ ) , I could easily make it to 100, but I find 10 should be enough. With different sized txt's, it is fine to tier them together without having to worry about a lack of randomness.

**Reply by Woisek** [reply]
Thanks for this fork, although I'm not quite sure how this would help with what I would like to have with Adetailer. 🙂
I basically just want to have Adetailer to respect variables that are set in the main prompt. 🤷‍♂️

**Comment by uorufu** [comment]
> Thanks for this fork, although I'm not quite sure how this would help with what I would like to have with Adetailer. 🙂 I basically just want to have Adetailer to respect variables that are set in the main prompt. 🤷‍♂️

Can you explain exactly what kind of variable you want to set?

The thing is that gradio is a bit of a pain when it comes to using info from older generations into newer ones (adetailer prompts are technically new generations). I fixed the adetailer issue for wildcards with some "random but not truly random" trickery. 

The `wildcard tier system` my fork has, can sort of pseudo set variables. If a wildcard in a particular text file that has 10 lines for example randomly chooses line 6. You can use another text file in the same tier and have something put on line 6 (and make sure that text file also have 10 lines) and then use that txt in adetailer without needing to put it in main prompt.

Thus technically having 'set a variable'.


Example:

If you have these two txt files

mainprompt.txt   prompt: __0_mainprompt__ 
a girl
a boy
a wolf

adprompt.txt  ad_prompt: __0_adprompt__ 
feminine
masculine
beastly

You can then use `__1_mainprompt__` in your normal prompt and `__1_adetailerprompt__` in your adetailer prompt and for every image that generates `a boy`, it will generate `masculine` in adetailer prompt and same for feminine and girl/wolf and beastly. As the tier line calculation results are linked as long as the txt files have same number of lines.
